homeassistant/components/enasolar/config_flow.py

```diff
@@ -0,0 +1,223 @@
+"""EnaSolar solar inverter configuration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import pyenasolar
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.core import HomeAssistant, callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    CONF_CAPABILITY,
+    CONF_DC_STRINGS,
+    CONF_HOST,
+    CONF_MAX_OUTPUT,
+    CONF_NAME,
+    CONF_SUN_DOWN,
+    CONF_SUN_UP,
+    DEFAULT_HOST,
+    DEFAULT_NAME,
+    DEFAULT_SUN_DOWN,
+    DEFAULT_SUN_UP,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def enasolar_entries(hass: HomeAssistant):
+    """Return the hosts already configured."""
+    return {
+        entry.data[CONF_HOST] for entry in hass.config_entries.async_entries(DOMAIN)
+    }
+
+
+class EnaSolarConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """User Configuration of EnaSolar Integration."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self._errors: dict[str, str] = {}
+        self._enasolar = None
+        self._data: dict[str, Any] = {}
+
+    def _host_in_configuration_exists(self, host) -> bool:
+        """Return True if host exists in configuration."""
+        if host in enasolar_entries(self.hass):
+            return True
+        return False
+
+    def _get_serial_no(self):
```

Comment:
I'm not sure that these functions need to be defined, you should just be able to call the `enasolar` method directly in the places where you are calling these one liners

Reply:
If I do that I cant work out how to mock out the call when doing the tests.

Reply:
I haven't written python before and I started by using another integration which, I now see, was probably not the most sensible thing to do as I didn't understand what all the code was doing.

Reply:
`homeassistant.components.enasolar.config_flow.EnaSolarConfigFlow._get_serial_no` becomes `homeassistant.components.enasolar.config_flow.pyenasolar.EnaSolar.get_serial_no`in your patch statement, rinse and repeat for the rest

Reply:
Moved device connection out of instantiation in pyenasolar which meant I could expose the methods directly and hence mock them out when testing.